sympy/matrices/expressions/matexpr.py

```diff
@@ -271,7 +272,7 @@ def I(self):
 
     def valid_index(self, i, j):
         def is_valid(idx):
-            return isinstance(idx, (int, Integer, Symbol, Expr))
+            return isinstance(idx, (int, Integer, numbers.Integral, Symbol, Expr))
```

Comment:
How about
```python
def is_valid(idx):
    try:
        idx = _sympify(idx)
        return isinstance(idx, (Integer, Symbol, Expr))
    except SympifyError:
        return False
```

Reply:
Both `Integer` and `Symbol` are subclasses of `Expr` so just `isinstance(idx, Expr)` should do it.

Reply:
I don't have a good way to check for `numbers` that are compatible with SymPy without converting though, it was the same thing I got stuck on when trying to improve `__eq__` for `Number` classes. Perhaps ABCs from `numbers` is the right approach, but that might not be stable enough (e.g. gmpy2 only being compatible in newer versions). It is also not clear to me how that would work precisely.

Reply:
Yes, I mean after sympifying.

Reply:
I got that, but ideally we would not need to convert `int` to `Integer` just to check if it is a valid index. Now putting in an extra instance check for `int` would be easy, but then why not also for `mpz` when it is available? (i.e. `SYMPY_INTS`). But if we go that route, we might indeed allow `numbers.Integral`. But that would possibly introduce a bug. I don't know if it will always work in that case.

Maybe the right approach is to check if it is an `Expr` and then try to convert it with `int(idx)`.

Reply:
Is `MatrixExpr.valid_index` intended to be public API? I agree that its only library use, from `MatrixExpr.__getitem__`, would not need the change to handle `numbers.Integral` because it is run on sympified data... but then neither would it need to test for `int`!
As it seems to be public, the change to allow `numbers.Integral` is an improvement.